ndcube/utils/wcs.py

```diff
@@ -140,75 +148,145 @@ def _wcs_slicer(wcs, missing_axis, item):
         Altered missing axis list.  Note the ordering has been reversed to reflect the data
         (numpy) axis ordering convention.
 
+    dropped_coords:
+        Coordinates which have been dropped in the slicing process is collected
+        in a dictionary called `dropped_coords`.
+
     """
+    # Force item to be WCS order if entered by user in numpy order.
+    if numpy_order is False:
```

Comment:
Based on above comment (which I think is right), this line should be
```python
if numpy_order is True:
```
Since ```numpy_order == True``` means the ```item``` is in the reverse order to the WCS.  But we want it to be the same order, as shown by our use of ```numpy_order=False``` in this [line](https://github.com/sunpy/ndcube/pull/151/files#diff-b324a6b383c9ce080bf07dcbacda10ddR273).  This may be causing some of the problems you've been seeing.

The idea is that the ```item``` is in numpy order for most of the slicing functions as that the same order as the data and how the user interacts with it.  However, because of annoying convention,  WCS order is reversed.  So in this function in particular we have the ```item``` entered in numpy order as that's where it is elsewhere (hence the ```numpy_order=True``` default in the function definition) and inside this function invert ```item``` to WCS order.  This makes it easier to interact with ```wcs``` and ```missing_axis```.  This also means when we call [wcs.slice](https://github.com/sunpy/ndcube/pull/151/files#diff-b324a6b383c9ce080bf07dcbacda10ddR273) we set ```numpy_order=False``` because we have already inverted the ```item```.  Does this makes sense?

Reply:
Sure, will change that soon... but this may introduce some changes to the code so I may need to debug further. Will take some time to check too.

Reply:
Would really like to seek your view on the following error though, since I have not encountered anything like it before. It has to do with one of the tests in `test_ndcube` and its traceback is as follows:
```
_________________ ERROR collecting ndcube/tests/test_ndcube.py _________________
ndcube/tests/test_ndcube.py:219: in <module>
    (cube[:, 1],
ndcube/mixins/ndslicing.py:36: in __getitem__
    kwargs, dropped_coords = self._slice(item)
ndcube/mixins/ndslicing.py:60: in _slice
    wcs, missing_axis, dropped_coords = self._slice_wcs_missing_axis(item)
ndcube/mixins/ndslicing.py:71: in _slice_wcs_missing_axis
    self.wcs, copy.deepcopy(self.missing_axis[::-1]), item)
ndcube/utils/wcs.py:263: in _wcs_slicer
    temp_real_world_coords = wcs.all_pix2world(*pix_coords, 0)[i]
/anaconda3/envs/ndcube-dev/lib/python3.7/site-packages/astropy/wcs/wcs.py:1297: in all_pix2world
    self._all_pix2world, 'output', *args, **kwargs)
/anaconda3/envs/ndcube-dev/lib/python3.7/site-packages/astropy/wcs/wcs.py:1286: in _array_converter
    return _return_list_of_arrays(axes, origin)
/anaconda3/envs/ndcube-dev/lib/python3.7/site-packages/astropy/wcs/wcs.py:1240: in _return_list_of_arrays
    output = func(xy, origin)
E   TypeError: Cannot cast array data from dtype('O') to dtype('float64') according to the rule 'safe'
```
I have no idea what the rule `'safe'` is. Could you tell me a bit more about it?

Reply:
Given that the `'O'` type is that of a Python object... I can in principle kind of force the conversion. Or I could try a gentler approach.

Reply:
Another representative error arising from the tests is as follows:
```
____________________________ ERROR collecting ndcube/tests/test_ndcubesequence.py _____________________________
ndcube/tests/test_ndcubesequence.py:45: in <module>
    cube1.extra_coords['pix']['value'][-1]),
E   TypeError: 'NoneType' object is not subscriptable
----------------------------------------------- Captured stdout -----------------------------------------------
[{'coordinate_type': None, 'scale': 'linear', 'group': 0, 'number': 0}, {'coordinate_type': 'spectral', 'scale': 'linear', 'group': 0, 'number': 0}, {'coordinate_type': 'celestial', 'scale': 'non-linear celestial', 'group': 0, 'number': 1}, {'coordinate_type': 'celestial', 'scale': 'non-linear celestial', 'group': 0, 'number': 0}]
```

Reply:
A good way to understand what errors mean is to insert print statements and rerun the test that should tell you what is making its way through the code and what the code is expecting to get.  Even then, the output of the tests can be very verbose.  Another way is to run the code line by line in your interpretter with the same inputs as the test.  Then you can inspect each variable at each stage and diagnose where the problem is.

Reply:
To be more specific about the first error, I don't know what the 'safe' rule is, but I don't think that's the key point.  The message tells me that the code is trying to convert an array to type float64.  However, it's receiving an array of objects which it can't safely convert.  Therefore the code isn't expected an object array so something has gone wrong, or at least changed.  So finding out what variable in what combination of inputs is the way forward.  The strategy I suggested above is probably the best one.